### PR TITLE
Ringbuf Support for Python API

### LINKED ISSUE #2985: Fully support new ringbuf map type in bcc
Linux 5.7 added a new ringbuf map type that can be used as an alternative to perf ring buffers for outputting custom events. I would be interested in having support for this in bcc.

I am willing to try to take a stab at adding support for this, but would need a bit of guidance first from a more experienced contributor/maintainer in terms of what changes would need to be made.

### PR DESCRIPTION
**Pull request by willfindlay**
This pull request contains an implementation for ringbuf support in bcc's Python API. 

Fixes #2985.

## Summary of Changes

- Added ringbuf helpers from libbpf API to libbcc
- Added a new `RingBuf` class to represent the ringbuf map
- Added a new `EventArrayBase` class to reduce code duplication between PerfEventArray and RingBuf
- Added `BPF_RINGBUF_OUTPUT` macro for BPF programs
- Added documentation

## Examples

The following example programs describe how the API works.

On the BPF side, we have two choices for submitting events. `map.ringbuf_output` works like `map.perf_submit` for perf buffers,
while `map.ringbuf_reserve`/`map.ringbuf_discard`/`map.ringbuf_submit` present an alternative approach that allows for finer control.

`map.ringbuf_output` works like:
```c
BPF_RINGBUF_OUTPUT(events);

struct data_t data {
    u32 pid;
};

TRACEPOINT_PROBE(raw_syscalls, sys_enter) {
    u32 pid = bpf_get_current_pid_tgid();

    struct data_t data = { .pid = pid };

    events.ringbuf_output(&data, sizeof(data), 0 /* flags */);
    
    return 0;
}
```

`map.ringbuf_reserve`/`map.ringbuf_discard`/`map.ringbuf_submit` work like:
```c
BPF_RINGBUF_OUTPUT(events);

struct data_t data {
    u32 pid;
};

TRACEPOINT_PROBE(raw_syscalls, sys_enter) {
    u32 pid = bpf_get_current_pid_tgid();

    struct data_t *data = events.ringbuf_reserve(sizeof(struct data_t));
    if (!data)
        return 1;

    data->pid = pid;

    events.ringbuf_submit(data, 0 /* flags */);
    // or, to discard the data...
    events.ringbuf_discard(data, 0 /* flags */);

    return 0;
}
```

On the Python side, the user calls `open_ring_buffer` to associate a callback with the buffer. There are two options for reading data from kernelspace: `ring_buffer_poll` which works like `perf_buffer_poll` and `ring_buffer_consume` which reads data without polling first.
```python
import time

from bcc import BPF

b = BPF(src_file=b'source.c')

def callback(ctx, data, size):
    event = b['events'].event(data)
    print(event.pid)

b['events'].open_ring_buffer(callback)

while True:
    # Option 1...
    b.ring_buffer_poll(30)
    # Option 2...
    b.ring_buffer_consume()
    time.sleep(0.5)
```

Unlike perf buffer callbacks, ringbuf supports return values from the callback, which can be used to indicate error conditions to stop polling early. If a callback does not return an integer, we fall back to returning 0, which approximates perf buffer behavior.

### CODE REVIEW

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by willfindlay** [comment]
Is there an easy way to see what tests specifically are failing? Seems to be something in `test_tools_smoke.py` from what I can tell.

**Comment by yonghong-song** [comment]
> Is there an easy way to see what tests specifically are failing? Seems to be something in `test_tools_smoke.py` from what I can tell.

Probably a flaky test. Let me try again.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by willfindlay** [comment]
Ah seems like it was indeed a flaky test.

File: src/python/bcc/table.py [hunk 1]
```diff
@@ -922,3 +947,48 @@ def __init__(self, *args, **kwargs):
...
+                    exit()
+                else:
+                    raise e
+            return ret
+
+        fn = _RINGBUF_CB_TYPE(ringbuf_cb_)
+        self.ringbuf = lib.bpf_new_ringbuf(self.map_fd, fn)
+        if not self.ringbuf:
+            raise Exception("Could not open ring buffer")
+        self.bpf.ring_buffers[id(self)] = self.ringbuf
```
Comment (anakryiko) [inline src/python/bcc/table.py:992]: libbpf's ring_buffer is already a manager of many ring buffers. It would be simpler to just use that, instead of maintaining list of ringbuffers on Python side? Unless I'm missing something, not an expert on BCC internals.

File: src/python/bcc/table.py [hunk 2]
```diff
@@ -922,3 +947,48 @@ def __init__(self, *args, **kwargs):
 class MapInMapHash(HashTable):
     def __init__(self, *args, **kwargs):
         super(MapInMapHash, self).__init__(*args, **kwargs)
+
+class RingBuf(EventArrayBase):
```
Comment (anakryiko) [inline src/python/bcc/table.py:951]: RingBuf has nothing in common with EventArrayBase, ringbuf is not an array, it doesn't share any functionality. I think this is just plain wrong and misleading.

File: src/cc/export/helpers.h [hunk 3]
```diff
@@ -141,6 +141,24 @@ struct _name##_table_t { \
 __attribute__((section("maps/perf_output"))) \
 struct _name##_table_t _name = { .max_entries = 0 }
 
+// Table for pushing custom events to userspace via ring buffer
+#define BPF_RINGBUF_OUTPUT(_name) \
```
[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/export/helpers.h:145]: I was experimenting with this earlier and it seems like most values cause map creation to fail with -EINVAL. Do you have some insight on why this might be the case?

File: src/python/bcc/table.py (hunk 1 above)
Comment (willfindlay) [reply src/python/bcc/table.py:992]: I chose to do it this way to make it more similar to perf buffer implementation. But I'm open to changing this.

File: src/python/bcc/table.py (hunk 2 above)
Comment (willfindlay) [reply src/python/bcc/table.py:951]: What would you rename it to? EventMapBase? EventTableBase?

File: src/cc/libbpf.h [hunk 4]
```diff
@@ -121,6 +121,18 @@ int bpf_open_perf_event(uint32_t type, uint64_t config, int pid, int cpu);
 
 int bpf_close_perf_event_fd(int fd);
 
+typedef int (*ring_buffer_sample_fn)(void *ctx, void *data, size_t size);
+
+struct ring;
```
[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/libbpf.h:126]: Good catch, I'll remove it.

File: src/cc/frontends/clang/b_frontend_action.cc [hunk 5]
```diff
@@ -1356,6 +1412,12 @@ bool BTypeVisitor::VisitVarDecl(VarDecl *Decl) {
       if (numcpu <= 0)
         numcpu = 1;
       table.max_entries = numcpu;
+    } else if (section_attr == "maps/ringbuf") {
+      map_type = BPF_MAP_TYPE_RINGBUF;
+      // values from libbpf/src/libbpf_probes.c
+      table.key_size = 0;
+      table.leaf_size = 0;
+      table.max_entries = 4096;
```
[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/frontends/clang/b_frontend_action.cc:1420]: This is the same place perf buffer does it.

File: src/python/bcc/table.py (hunk 2 above)
Comment (willfindlay) [reply src/python/bcc/table.py:951]: Or maybe EventBufferBase?

File: src/cc/libbpf.c [hunk 6]
```diff
@@ -1399,3 +1399,36 @@ int bpf_close_perf_event_fd(int fd) {
   }
   return error;
 }
+
+/* Create a new ringbuf manager to manage ringbuf associated with
+ * map_fd, associating it with callback sample_cb. */
+void * bpf_new_ringbuf(int map_fd, ring_buffer_sample_fn sample_cb) {
+    //return ring_buffer__new(map_fd, sample_cb, (void *)(long)1, NULL);
+    return ring_buffer__new(map_fd, sample_cb, NULL, NULL);
```
[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/libbpf.c:1407]: I can change this. Do you mind going over one such use case? When I was looking at examples in Linux selftests I couldn't quite work out what ctx was used for.

**Comment by willfindlay** [comment]
Thank you for your comments @anakryiko. I left a few follow up questions / concerns as replies. I should have time to update this PR later today or tomorrow.

File: src/cc/export/helpers.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/export/helpers.h]: I've added a commit that allows the user to specify max_entries by number of pages.

File: src/cc/libbpf.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (anakryiko) [reply src/cc/libbpf.c:1407]: look at ring_buffer_sample_fb signature. First argument is `void *ctx`. That's the same pointer that's passed as third argument to ring_buffer__new. Also each ring_buffer__add() allows to add another ring_buffer map to one common "ring buffer manager", and it also takes this context, which will be passed to callback for samples from that particular ring buffer.

As for how it's used, it's very common to pass a pointer to an object managing data consumption (in C++, especially). That way callback itself is a global static function, but can work with non-global state.

File: src/python/bcc/table.py (hunk 1 above)
Comment (anakryiko) [reply src/python/bcc/table.py:992]: I specifically designed ring_buffer in libbpf to be able to handle multiple ring buffer maps simultaneously within a single epoll instance. I saw that in practice some applications have multiple perfbuf/ringbufs and they are going in loop to poll them, which is very suboptimal. So while we don't have much choice for perf_buffer, please do better for ring_buffers.

File: src/python/bcc/table.py (hunk 2 above)
Comment (anakryiko) [reply src/python/bcc/table.py:951]: I don't know if we need a common Base? What's to be shared in terms of functionality?

File: src/cc/export/helpers.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (anakryiko) [reply src/cc/export/helpers.h]: yes, it has to be in multiples of pages. I'd stick to bytes, though, because that's what kernel expects. If you want to be helpful to users (at the expense of keeping them in the dark about what's actually happening), you can automatically round-up to the closest multiple of page_size.

[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/export/helpers.h]: @anakryiko What I'm doing right now is something like `BPF_RING_OUTPUT(buffer, 8)` which results in `{.max_entries = (8 * PAGE_SIZE)}`. I think this should be fine as an interface, as this resembles what we do for `PERF_OUTPUT` on the python side (user specifies number of pages when calling `open_perf_buffer`).

[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/export/helpers.h]: So the kernel still sees bytes, but user simply provides number of pages.

File: src/cc/libbpf.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/libbpf.c:1407]: I see. I suppose it doesn't make a lot of difference from the Python side, but I will certainly modify libbcc to support this. Thanks for the clarification.

File: src/python/bcc/table.py (hunk 2 above)
Comment (willfindlay) [reply src/python/bcc/table.py:951]: @anakryiko The common base is for resolving event types. See the methods exposed by EventArrayBase. Before this was all part of PerfEventArray, but keeping it that way would result in code duplication.

File: src/python/bcc/table.py (hunk 1 above)
Comment (willfindlay) [reply src/python/bcc/table.py:992]: The performance argument makes a lot of sense. This will represent a significant change compared to how we handle other map types, but it is probably for the best. I'll try and get this working as soon as I can.

File: src/cc/export/helpers.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (anakryiko) [reply src/cc/export/helpers.h]: I'm worried that it just deviates from what's happening when people use libbpf instead of BCC and would complicate, unnecessarily, migration process.

File: src/python/bcc/table.py (hunk 2 above)
Comment (anakryiko) [reply src/python/bcc/table.py:951]: You mean _get_event_class() stuff? Couldn't that be just a static helper method (or whatever Python equivalent of that)? EventArrayBase doesn't make any sense whatsoever for ringbuf, because it neither is an array-like map, nor it deals with perf events.

File: src/cc/export/helpers.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/export/helpers.h]: Not sure that I agree with this. I feel like it's simple enough to change to bytes when switching to use libbpf, and using bytes instead of pages here would not only confuse new users (when they can't figure out why their maps aren't working), but would also represent a significant departure for people who are already familiar with perf buffers in bcc, which deal in pages.

[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/export/helpers.h]: @yonghong-song Do you have any thoughts on this?

[the remark this answers is not in the document]
Comment (anakryiko) [reply src/cc/export/helpers.h]: I don't feel very strongly on this, so feel free to stick to page count. But I'd love users to at least give it a passing thought that in case of PERFBUF it's pages of data per-CPU vs for RINGBUF it's one common ringbuf.

File: src/python/bcc/table.py (hunk 1 above)
Comment (anakryiko) [reply src/python/bcc/table.py:992]: Latency as well, not just performance :)

File: src/python/bcc/table.py (hunk 2 above)
Comment (willfindlay) [reply src/python/bcc/table.py:951]: @anakryiko Fair enough, that sounds like a good solution.

File: src/cc/export/helpers.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (willfindlay) [reply src/cc/export/helpers.h]: I'll be certain to mention this when I write the documentation.

**Comment by willfindlay** [comment]
@anakryiko I have now addressed all of your feedback (with the exception of pages vs bytes in the ringbuf macro, as discussed). Things seem to be in a good place now, but if you have any other concerns let me know.

**Comment by willfindlay** [comment]
@yonghong-song Should be good to test again.

**Comment by anakryiko** [comment]
@yonghong-song, do you squash all those 44 commits into one when you land? I wonder if it makes sense to ask people to re-organize and squash into fewer logic commits before landing. E.g., implementation, testing and documentation commits. 44 commits are pretty unwieldy.

**Comment by yonghong-song** [comment]
@anakryiko Thanks for review. I won't push these 44 commits. I indeed frequently squash them into one commit. llvm sometimes also has giant commits. I think it is okay not to put burden on the developer for already done work.
I will take a look shortly.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by yonghong-song** [comment]
[buildbot, test this please]

File: docs/reference_guide.md [hunk 7]
```diff
@@ -1694,6 +1876,68 @@ Examples in situ:
 [search /examples](https://github.com/iovisor/bcc/search?q=print_linear_hist+path%3Aexamples+language%3Apython&type=Code),
 [search /tools](https://github.com/iovisor/bcc/search?q=print_linear_hist+path%3Atools+language%3Apython&type=Code)
 
+### 8. open_ring_buffer()
+
+Syntax: ```table.open_ring_buffer(callback, ctx=None)```
+
+This operates on a table as defined in BPF as BPF_RINGBUF_OUTPUT(), and associates the callback Python function ```callback``` to be called when data is available in the ringbuf ring buffer. This is part of the new (Linux 5.8+) recommended mechanism for transferring per-event data from kernel to user space. Unlike perf buffers, ringbuf sizes are specified within the BPF program, as part of the ```BPF_RINGBUF_OUTPUT``` macro. If the callback is not processing data fast enough, some submitted data may be lost. In this case, the events should be polled more frequently and/or the size of the ring buffer should be increased.
```
[the remark this answers is not in the document]
Comment (willfindlay) [reply docs/reference_guide.md:1883]: I don't think this would be easy to implement given the way bcc treats maps. We could potentially change this in the future, but for now, I think the current API is fine. For now, if the user desires different map sizes according to userspace logic, they can just programmatically generate the correct C code. This is already a common pattern in bcc tools.

**Comment by yonghong-song** [comment]
Merged. Thanks @willfindlay and @anakryiko!

**Comment by raoufkh** [comment]
Hello

First, thank you @willfindlay  for this work. 

I have one question, in TC and XDP programs, how can we send the network packet to the user space program. Using perf buffers, this was possible by using the `perf_submit_skb` method while `perf_submit` allows only a data struct we define.

I'm trying to include an `unsigned char *` pointer in my struct (which I will send to the user space program using `ringbuf_output` but I don't manage to copy the network packet inside the struct. Doing so, the struct memory space will contain the value of the pointer (memory address of netwokr packet) but not the network packet itself. Regarding to the linux documentation, ring buffers should allow sending entire network packet to the user space. 

Could you provide help please?

Regards
Raouf
